### ISSUE TITLE
Neural Collaborative Filtering

### ISSUE THREAD

**Issue by jkwag** [AIA-AI-Implementation-Archive/aia#8]
# 논문 링크

https://arxiv.org/pdf/1708.05031

# 모델 설명

Neural Collaborative Filtering(2017)의 경우 추천 시스템 도메인에서 deep neural network 를 차용하려는 시도는 여러개가 있었으나 item의 textual description 과 같이 부가적인 정보만 활용하는 한계점을 지적하며, Collaborative Filtering의 핵심 factor 인 user 와 item의 interaction을 기존의 MF를 활용해 표현하는 것을 NCF 라는 neural architecture 기반의 framework를 제시했다. 

## MF의 한계

Matrix Factorization은 Collaborative Filtering Technique들 중 가장 인기가 많은 방식으로 user와 item을 공통의 latent space로 project 하여 inner product를 활용해 user의 item에 대한 선호도를 모델링한다. 다만, inner product 방식은 latent feature들의 단순 곱셈, 즉 선형 계산에 불과하기 때문에 user interaction data의 복잡한 구조를 반영하지 못한다. 

![image](https://github.com/user-attachments/assets/9ea4b5a7-23f6-4cc7-a52c-451a986a0e64)

위의 이미지에서 선형 계산의 한계가 두드러지게 나타난다. (a) 같은 경우 유저와 아이템의 implicit data 를 interaction matrix로 나타낸 것이고 (b)는 (a)의 matrix를 통해 2차원 공간으로 user를 latent space로 project 시켰을 때 어떤 방식으로 배치될지에 대한 예시이다. 

각각의 행과 열에 위치한 $u_{i}$ 와 $i_{j}$ 는 user와 item을 나타내는데 (a)에서는 $u_{1}, u_{2}, u_{3}$이 있다고 가정하고 item 의 경우 $i_{1}$ 부터 $i_{5}$ 까지 있는데 여기에 $u_{4}$가 추가되는 걸 가정하는 상황이다. $u_{4}$는 $u_{1}$과 제일 흡사한 취향을 가지고 있어 (b)와 같이 latent space상에서는 $p_{1}$에 가장 가까이 위치하면서 왼쪽이나 오른쪽에 배치됨이 바람직하다.

하지만, (a)를 참조한다면 $u_{2}$의 경우 $u_{1}$ 과 취향이 $u_{3}$와 비교해 더 가깝고 따라서 $p_{2}$ 가 $p_{1}$ 과 $p_{3}$ 의 중간에 위치한다. $u_{4}$의 경우 $u_{3}$와 취향이 더 흡사하므로 (b) latent space 위치 상 $p_{4}$ 는 $p_{2}$에 $p_{3}$ 보다 더 가깝게 위치하고 따라서 큰 ranking loss에 기여하게 된다. 

위의 간단한 예시는 복잡한 user-item interaction을 저차원의 latent space상에서 단순한 inner product로 나타내기에는 한계점이 명확한 점을 보여주고 있다. 이를 해결하기 위한 방법 중 하나는 latent space 공간을 크게(latent factor 의 수를 늘리는) 만드는 방법이 있으나 이는 과적합의 위험이 있고, 특히나 추천 시스템은 sparse한 데이터가 대부분이므로 과적합의 위험이 부각된다.

## Neural Collaborative Filtering

따라서, 비선형성을 반영하는 Deep Neural Network 구조를 collaborative filtering에 사용하는 것을 논문의 저자는 주장하였고 이를 NCF라 명명하고 MF 또한 NCF의 한 종류로 나타낼 수 있다고 주장했다. 또한, NCF framework 기반 MF 와 MLP를 앙상블한 새로운 MF 모델을 소개했는데 이 모델이 구현 프로젝트의 목표이다. 
![image](https://github.com/user-attachments/assets/13fcf856-4f72-442e-94c9-e110dc992c8f)

![image](https://github.com/user-attachments/assets/a2958741-d284-4a30-9294-f7122be6db72)

NCF의 예측모델 공식은 위와 같다. 예측 평점 $\hat{y_{ui}}$ 은 latent factor matrix인 P 와 Q의 user u 와 item i 를 나타내는 피쳐 벡터인 $v_{u}$ 와 $v_{I}$에 각각 내적된 값이 f를 통해 최종적으로 산출값이 나오는데 이 두 피쳐 벡터들에는 논문에서는 유저와 아이템 id 만을 사용했지만 context aware 하거나 content-based한 모델에 사용되는 피쳐를 모델링하는 벡터들을 차용할 수 있다. 
$\theta_{f}$ 는 식에서 interaction function f의 모델 파라미터를 나타낸다. f는 multi-layer neural network이기 때문에

![image](https://github.com/user-attachments/assets/826fd550-ee20-47e1-8120-92d612661a72)

다음 식과 같이 나타낼 수 있고, $\phi_{out}$과 $\phi_{x}$는 각각 output layer와 x번째 CF layer의 mapping function이다. 

![image](https://github.com/user-attachments/assets/b4501819-2182-4d42-b2cd-9cf98cad291d)

P와 Q는 latent factor matrix이기 때문에 다음과 같은 차원을 가진다.

## NCF framework 에서 MF(GMF)

MF 는 NCF의 special case중 하나로 볼 수 있는데 이유는 다음과 같다. input layer에서 embedding vector를 가져오는 과정은 latent space로 project하는 과정으로 볼 수 있다. 즉, MF 논문에서 언급되는 user latent vector $p_{u}$와 item latent vector $q_{i}$ 는 $P^{T}v_{u}$ 와 $Q^{T}v_{i}$ 로 볼 수 있고 첫 번쨰 neural CF layer는 두 latent vector 의 내적 그리고, 마지막 출력층에서

![image](https://github.com/user-attachments/assets/3ae94b41-4f30-4661-b903-cddbb3cf8595)

$alpha_{out}$은 identity function으로 hidden function인 $h$는 1로 채워진 uniform vector로 볼 수 있어 NCF framework 내에서 MF 모델을 나타낼 수 있다. 논문에서는 $alpha_{out}$ function을 sigmoid로 $h$를 log loss 를 통해서 데이터로부터 학습하는 방식을 차용하여 이를 GMF(Generalized Matrix Factorization)이라 명명했다.

## NCF framework에서 MLP

NCF에서 user와 item을 각각의 방법으로 모델링 하기 때문에 둘을 concatenate 하는것이 multimodal deep learning 에서 널리 차용되었듯이 자연스러운 방법이다. 하지만, 둘을 단순히 concatenate 하는건 둘의 interaction effect를 충분히 반영하지 못한다 설명하며 논문의 저자는 concatenate 된 벡터에 hidden layer를 추가하는 것이 바람직하다고 설명한다. 

![image](https://github.com/user-attachments/assets/6c3da4f6-721d-417f-ae42-8d859d4e0867)

따라서, 처음에 concatenate 하는 과정을 거치고 $W$ 로 대표되는 weight matrix, $b$ 로 대표되는 bias vector, $alpha$ 로 대표되는 activation function 을 여러층으로 쌓아 최종적으로 sigmoid 함수를 통해 $\hat y_{ui}$ 을 산출해낸다. Activation function 같은 경우 sigmoid, tank, RELU 같은 함수를 자유롭게 선정할 수 있는데 sparse data에 적합한 RELU가 권장된다.

## GMF와 MLP의 통합

![image](https://github.com/user-attachments/assets/7ade820e-b39f-4070-a2db-b8709b897ed6)

NCF Framework내에서 GMF 와 MLP의 예시들을 들었으므로 둘의 강점을 이용할 수 있는 구조에 대한 고민이 따르는데 가장 직관적인 방법으로는 임베딩 레이어를 공유하게 하고 interaction function들의 층들을 거쳐 output만 결합하는 것이다. 하지만, 임베딩 레이어를 공유한다는 건 같은 차원(크기)의 임베딩을 공유해야 된다는 제한이 따른다는 것이고 다양한 데이터셋에서 각기 다른 GMF와 MLP의 이상적인 임베딩 크기가 다를 수 있으므로 효과적인 앙상블이 나오기 어렵다.

따라서, 논문의 저자들은 통합된 모델의 flexibility를 위해 임베딩 또한 다르게 학습하고 두 모델을 마지막 hidden layer를 통해 통합하는 모델을 제시하였고 이를 NeuMF 라고 명명하였다. MF의 선형적인 단순하지만 강력한 user-item inner product 와 MLP 의 비선형적인 activation function을 통해 복잡한 관계를 표현하면서 둘의 강점을 살릴 수 있다.

### Pre-training

NeuMF는 back-propagation을 통해 model parameter를 업데이트 하는데 objective function 이 non convex 하기 때문에 global optimal 한 해를 찾는 것이 보장되지 않고 따라서, initialization이 converge 하는데 굉장히 중요한 역할을 한다. 또한, NeuMF는 GMF와 MLP의 앙상블이기 때문에 GMF와 MLP를 각각 pretraining 시킴으로서 이 문제를 해결할 수 있다. 

먼저, GMF와 MLP를 converge 할 때까지 random initialization 후 각각 train을 시킨다. 이후, output layer만 tweak되는데 마지막 output layer는 다음과 같다. 

![image](https://github.com/user-attachments/assets/76816f65-84c0-4018-83f6-97df3116908a)

$h^{GMF}$와 $h^{MLP}$는 pre-train된 GMF와 MLP 벡터를 나타내고 $alpha$는 두 모델의 비중을 조절하는 hyperparameter이다. 두 모델을 각각 training 을 할 때 Adam이 사용되는데 learning rate 을 튜닝하는 수고가 덜어지고 더 빠르게 converge했기 때문에 SGD 대신 사용됐다고 논문의 저자들을 설명했다.  

# 논문 선정 이유

추천 시스템에서 딥러닝 기반의 모델들이 상대적으로 덜 연구되었는데 신호탄을 쐈던 논문이였고 NeuMF는 MF와 MLP라는 Collaborative Filtering과 Deep learning을 받치는 초석의 모델들이 결합됐기에 이를 토대로 발전된 모델들을 적용시킬 수 있는 근본적인 모델을 먼저 구현해 보고 싶은 동기부여가 있었고 구현하면서 자유롭게 구조를 바꾸는 실험 같은 것을 진행해볼 수도 있다.

# 논문 구현 예상 기간

2주정도 예상
